### PR TITLE
MULE-19916: Redelivery Policy: failure to obtain message ID from expression causes source to stop listening for messages when it uses transactions.

### PR DESCRIPTION
**Pull request by rbourbonnavarro**

### CODE REVIEW

File: integration/src/test/resources/org/mule/test/components/redelivery-policy-config.xml [hunk 1]
```diff
@@ -95,4 +114,35 @@
...
+    <flow name="redeliveryInvalidMessageIdWithTransactionalSourceAndCustomErrorHandlerProcess">
+        <vm:listener queueName="Q5" config-ref="VM_Config" transactionalAction="ALWAYS_BEGIN">
+            <redelivery-policy idExpression="payload.id ++ '-' ++ payload.id"/>
+        </vm:listener>
+
+        <logger/>
+
+        <error-handler>
+            <on-error-propagate type="EXPRESSION">
+                <set-payload value="#[error]"/>
```
Comment (elrodro83) [inline integration/src/test/resources/org/mule/test/components/redelivery-policy-config.xml:131]: use `#[!null.a]` expression to force an expression error

File: integration/src/test/resources/org/mule/test/components/redelivery-policy-config.xml [hunk 2]
```diff
@@ -9,6 +9,15 @@
                            http://www.mulesoft.org/schema/mule/vm http://www.mulesoft.org/schema/mule/vm/current/mule-vm.xsd
                            http://www.mulesoft.org/schema/mule/test-components http://www.mulesoft.org/schema/mule/test-components/current/mule-test-components.xsd">
 
+    <configuration defaultErrorHandler-ref="globalErrorHandler" />
+
+    <error-handler name="globalErrorHandler">
+        <on-error-propagate type="EXPRESSION">
+            <set-payload value="#[error]"/>
```
Comment (elrodro83) [inline integration/src/test/resources/org/mule/test/components/redelivery-policy-config.xml:16]: use `#[!null.a]` expression to force an expression error

**Review by elrodro83 — CHANGES_REQUESTED** [review]

File: integration/src/test/resources/org/mule/test/components/redelivery-policy-config.xml (hunk 1 above)
Comment (rbourbonnavarro) [reply integration/src/test/resources/org/mule/test/components/redelivery-policy-config.xml:131]: I didn't intend to force an expression error there, just get the error itself for the log. However I did change it in the two places where I needed to force an expression error.

File: integration/src/test/resources/org/mule/test/components/redelivery-policy-config.xml (hunk 2 above)
Comment (rbourbonnavarro) [reply integration/src/test/resources/org/mule/test/components/redelivery-policy-config.xml:16]: I didn't intend to force an expression error there, just get the error itself for the log. However, I did change it in the two places where I needed to force an expression error.

File: integration/src/test/resources/org/mule/test/components/redelivery-policy-config.xml [hunk 3]
```diff
@@ -95,4 +120,43 @@
...
+        </error-handler>
+    </flow>
+
+    <flow name="redeliveryInvalidMessageIdWithTransactionalSourceAndDefaultErrorHandlerDispatch">
+        <vm:publish-consume queueName="Q6" config-ref="VM_Config"/>
+    </flow>
+
+    <flow name="redeliveryInvalidMessageIdWithTransactionalSourceAndDefaultErrorHandlerProcess">
+        <vm:listener queueName="Q6" config-ref="VM_Config" transactionalAction="ALWAYS_BEGIN">
+            <redelivery-policy idExpression="!null.a"/>
```
Comment (elrodro83) [inline integration/src/test/resources/org/mule/test/components/redelivery-policy-config.xml:149]: is this of without the `#[]` marker?

File: integration/src/test/resources/org/mule/test/components/redelivery-policy-config.xml [hunk 4]
```diff
@@ -95,4 +120,43 @@
...
+        </vm:listener>
+
+        <logger/>
+    </flow>
+
+    <flow name="redeliveryInvalidMessageIdWithHttpListener">
+        <http:listener path="invalidMessageId" config-ref="listenerConfig">
+            <redelivery-policy idExpression="!null.a"/>
```
Comment (elrodro83) [inline integration/src/test/resources/org/mule/test/components/redelivery-policy-config.xml:157]: is this of without the `#[]` marker?

File: integration/src/test/resources/org/mule/test/components/redelivery-policy-config.xml [hunk 5]
```diff
@@ -95,4 +120,43 @@
...
         </error-handler>
     </flow>
+
+    <flow name="redeliveryInvalidMessageIdWithTransactionalSourceAndCustomErrorHandlerDispatch">
+        <vm:publish-consume queueName="Q5" config-ref="VM_Config"/>
+    </flow>
+
+    <flow name="redeliveryInvalidMessageIdWithTransactionalSourceAndCustomErrorHandlerProcess">
+        <vm:listener queueName="Q5" config-ref="VM_Config" transactionalAction="ALWAYS_BEGIN">
+            <redelivery-policy idExpression="!null.a"/>
```
Comment (elrodro83) [inline integration/src/test/resources/org/mule/test/components/redelivery-policy-config.xml:130]: is this of without the `#[]` marker?

**Review by elrodro83 — CHANGES_REQUESTED** [review]

File: integration/src/test/resources/org/mule/test/components/redelivery-policy-config.xml (hunk 3 above)
Comment (rbourbonnavarro) [reply integration/src/test/resources/org/mule/test/components/redelivery-policy-config.xml:149]: It actually works seamlessly with or without the `#[]` expression marker.

File: integration/src/test/resources/org/mule/test/components/redelivery-policy-config.xml (hunk 4 above)
Comment (rbourbonnavarro) [reply integration/src/test/resources/org/mule/test/components/redelivery-policy-config.xml:157]: It actually works seamlessly with or without the `#[]` expression marker.

File: integration/src/test/resources/org/mule/test/components/redelivery-policy-config.xml (hunk 5 above)
Comment (rbourbonnavarro) [reply integration/src/test/resources/org/mule/test/components/redelivery-policy-config.xml:130]: It actually works seamlessly with or without the `#[]` expression marker.

**Comment by rbourbonnavarro** [comment]
--validate

**Review by elrodro83 — APPROVED** [review]

**Pull request merged**